sdk/core/Azure.Core/src/Pipeline/Internal/RetryPolicy.cs

```diff
@@ -20,46 +23,59 @@ internal class RetryPolicy : HttpPipelinePolicy
 
         private readonly Random _random = new ThreadSafeRandom();
 
-        public RetryPolicy(RetryMode mode, TimeSpan delay, TimeSpan maxDelay, int maxRetries)
-        {
-            _mode = mode;
-            _delay = delay;
-            _maxDelay = maxDelay;
-            _maxRetries = maxRetries;
-        }
-
         private const string RetryAfterHeaderName = "Retry-After";
         private const string RetryAfterMsHeaderName = "retry-after-ms";
         private const string XRetryAfterMsHeaderName = "x-ms-retry-after-ms";
 
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="options"></param>
+        public RetryPolicy(RetryOptions options)
+        {
+            _mode = options.Mode;
+            _delay = options.Delay;
+            _maxDelay = options.MaxDelay;
+            _maxRetries = options.MaxRetries;
+        }
+
+        /// <inheritdoc/>
         public override void Process(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
         {
             ProcessAsync(message, pipeline, false).EnsureCompleted();
         }
 
+        /// <inheritdoc/>
         public override ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
         {
             return ProcessAsync(message, pipeline, true);
         }
 
         private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
         {
-            int attempt = 0;
+            message.RetryContext = new RetryContext(DateTimeOffset.UtcNow);
             List<Exception>? exceptions = null;
             while (true)
             {
-                Exception? lastException = null;
+                message.RetryContext.AttemptNumber++;
                 var before = Stopwatch.GetTimestamp();
                 try
                 {
                     if (async)
                     {
+                        await OnTryRequestAsync(message).ConfigureAwait(false);
                         await ProcessNextAsync(message, pipeline).ConfigureAwait(false);
+                        await OnResponseAsync(message).ConfigureAwait(false);
                     }
                     else
                     {
+                        OnTryRequest(message);
```

Comment:
We probably don't want to catch user exceptions in OnTryRequest/OnResponse. Instead we can move these outside of the try.